src/Libraries/Microsoft.Extensions.Telemetry.Testing/Metering/MetricCollector.cs

```diff
@@ -2,446 +2,540 @@
+    /// <param name="timeProvider">The time provider against which the timeout executes. If this is <see langword="null"/> then the provider
+    /// selected when creating the metric collector is used.</param>
+    /// <returns>A task that completes when the collector has collected the requisite number of measurements.</returns>
+    [SuppressMessage("Resilience", "R9A061:The async method doesn't support cancellation", Justification = "Not relevant in this case")]
+    public Task WaitForMeasurementsAsync(int numMeasurements, TimeSpan timeout, TimeProvider? timeProvider = null)
+    {
+        _ = Throw.IfLessThan(numMeasurements, 1);
+        _ = Throw.IfLessThan(timeout.Ticks, 0);
 
-    /// <summary>
-    /// Gets a list of all observable counters registered with this metrics collector.
-    /// </summary>
-    /// <typeparam name="T">The type of metric measurement value.</typeparam>
-    /// <returns>Read only dictionary of <see cref="MetricValuesHolder{T}"/>.</returns>
-    public IReadOnlyDictionary<string, MetricValuesHolder<T>>? GetAllObservableCounters<T>()
-        where T : struct => GetAllInstruments<T>(_allObservableCounters);
+        lock (_measurements)
+        {
+            ThrowIfDisposed();
 
-    /// <summary>
-    /// Gets a list of all observable UpDown counters registered with this metrics collector.
-    /// </summary>
-    /// <typeparam name="T">The type of metric measurement value.</typeparam>
-    /// <returns>Read only dictionary of <see cref="MetricValuesHolder{T}"/>.</returns>
-    public IReadOnlyDictionary<string, MetricValuesHolder<T>>? GetAllObservableUpDownCounters<T>()
-        where T : struct => GetAllInstruments<T>(_allObservableUpDownCounters);
+            if (_measurements.Count >= numMeasurements)
+            {
+                return Task.CompletedTask;
+            }
 
-    /// <summary>
-    /// Gets a list of all observable Gauges registered with this metrics collector.
-    /// </summary>
-    /// <typeparam name="T">The type of metric measurement value.</typeparam>
-    /// <returns>Read only dictionary of <see cref="MetricValuesHolder{T}"/>.</returns>
-    public IReadOnlyDictionary<string, MetricValuesHolder<T>>? GetAllObservableGauges<T>()
-        where T : struct => GetAllInstruments<T>(_allObservableGauges);
-#pragma warning restore S4049 // Properties should be preferred
+            var w = new Waiter(numMeasurements);
+            _waiters.Add(w);
+
+            return w.TaskSource.Task.WaitAsync(timeout, timeProvider ?? _timeProvider, CancellationToken.None);
+        }
+    }
 
     /// <summary>
-    /// Collects all observable instruments and records their measurements.
+    /// Scan all registered recorded instruments.
     /// </summary>
-    public void CollectObservableInstruments()
+    public void RecordObservableInstruments()
     {
-        _listener.RecordObservableInstruments();
+        if (_instrument is ObservableInstrument<T>)
```

Comment:
Nit: I don't think this conditional is necessary, MeterListener.RecordObservableInstruments() should already be a no-op in this case.